ast_tools/passes/ssa.py

```diff
@@ -470,9 +470,21 @@ def rewrite(self,
         # before any transformation happens
         NR = _never_returns(tree.body)
 
+        replacer = AttrReplacer(node_table={})
+
+        # Find all self.attr.prev() call
+        calls = collect_calls(tree)
+        for c in calls:
+            if (isinstance(c.func, ast.Attribute) and
+                    isinstance(c.func.value, ast.Attribute) and
+                    isinstance(c.func.value.value, ast.Name) and
+                    c.func.value.value.id == "self" and
+                    c.func.attr == "prev"):
```

Comment:
This needs to be factored into extensible logic.   I am very opposed to special casing random attribute names.

What is the transformation you are trying to accomplishment?  Why not put it in a separate pass?  If that is not feasible can we make this some sort of callback / argument to `__init__`?

Reply:
I think this would make more sense to be a magma preprocessing pass and it's specific to sequential/coroutine (it should be done inside sequential2) or perhaps we can find some other way to support it without having to run a pass on sequential (what we were hoping with sequential2), but I think this is too specific for ssa (the goal is for ssa to be a generally useful pass, not magma specific).

Reply:
If this does something that might have a general use case I am not opposed to adding a call back if there is no way to achieve this in a separate pass. I just need to understand the goals / requirements.

Reply:
sequential has the ability to say
```
self.reg += 1
return self.reg.prev()
```
where this will return the previous value of self.reg, even though it's been updated in the call body.

The motivation is that it's common for circuits to have registered outputs, so this is syntax sugar for storing the previous value of self.reg at the top of the function, and returning that value.  It's a convenience method to cover this common use case and is really magma specific I would say.

Reply:
Hmm, I think the real issue is I don't support method calls on written attributes.  

```Python
__return_value0 = self.cnt
```

Should be 

```Python
__return_value0 = self_cnt1.prev()
```

I don't see *this* as a magma specific issue.  Although the current solution definitely is.  I am also a little surprised that this wasn't generated.  I look into it.

Reply:
Hmm, well if we can support that rewrite, then perhaps we can add some magic to the 'prev()' method to make this work without having to do a special rewrite in magma

Reply:
(I'll have to double check the old sequential code, but I'm pretty sure we just rewrote these references during the rewrite pass, so we didn't use any magic method)

Reply:
Just tried the test that @splhack wrote on master.  It worked fine.  It also worked on `v0.0.17` so it wasn't the recent commit.  

@splhack do you think you can generate a minimal failing test case and put in an issue?